tensornetwork/contractors/opt_einsum_paths/path_contractors.py

```diff
@@ -0,0 +1,158 @@
+    memory_limit: Maximum number of elements in an array during contractions.
+
+  Returns:
+    The network after full contraction.
+  """
+  alg = functools.partial(opt_einsum.paths.optimal, memory_limit=memory_limit)
+  return base(net, alg)
+
+
+def branch(net: network.TensorNetwork, memory_limit: Optional[int] = None,
+           nbranch: Optional[int] = None) -> network.TensorNetwork:
+  """Branch contraction path via `opt_einsum`.
+
+  This method uses the DFS approach of `optimal` while sorting potential
+  contractions based on a heuristic cost, in order to reduce time spent
+  in exploring paths which are unlikely to be optimal.
+  For more details:
+    https://optimized-einsum.readthedocs.io/en/latest/branching_path.html
+
+  Args:
+    net: a TensorNetwork object.
+    memory_limit: Maximum number of elements in an array during contractions.
+    nbranch: Number of best contractions to explore.
+      If None it explores all inner products starting with those that
+      have the best cost heuristic.
+
+  Returns:
+    The network after full contraction.
+  """
+  alg = functools.partial(opt_einsum.paths.branch, memory_limit=memory_limit,
+                          nbranch=nbranch)
+  return base(net, alg)
+
+
+def greedy(net: network.TensorNetwork,
+           memory_limit: Optional[int] = None) -> network.TensorNetwork:
+  """Greedy contraction path via `opt_einsum`.
+
+  This provides a more efficient strategy than `optimal` for finding
+  contraction paths in large networks. First contracts pairs of tensors
+  by finding the pair with the lowest cost at each step. Then it performs
+  the outer products.
+  For more details:
+    https://optimized-einsum.readthedocs.io/en/latest/greedy_path.html
+
+  Args:
+    net: a TensorNetwork object.
+    memory_limit: Maximum number of elements in an array during contractions.
+
+  Returns:
+    The network after full contraction.
+  """
+  alg = functools.partial(opt_einsum.paths.greedy, memory_limit=memory_limit)
+  return base(net, alg)
+
+
+def auto(net: network.TensorNetwork,
+         memory_limit: Optional[int] = None) -> network.TensorNetwork:
+  """Chooses one of the above algorithms according to network size.
+
+  Default behavior is based on `opt_einsum`'s `auto` contractor.
+
+  Args:
+    net: a TensorNetwork object.
+    memory_limit: Maximum number of elements in an array during contractions.
+
+  Returns:
+    The network after full contraction.
+  """
+  n = len(net.nodes_set)
+  if n <= 0:
+    raise ValueError("Cannot contract empty tensor network.")
+  if n == 1:
+    edges = net.get_all_nondangling()
+    net.contract_parallel(edges.pop())
+    return net
+  if n < 5:
+    return optimal(net, memory_limit)
+  if n < 7:
+    return branch(net, memory_limit)
+  if n < 9:
+    return branch(net, memory_limit, nbranch=2)
+  if n < 15:
+    return branch(net, nbranch=1)
+  return greedy(net, memory_limit)
+
+
+def custom(net: network.TensorNetwork, optimizer,
```

Comment:
add typing to optimizer. 

(Last comment. I promise :))

Reply:
No worries! :D I just hope tests pass because I had it and removed it for this reason (I think I had typo)